### PR TITLE
Add subcommand to "hydrate" a task dump

### PR DESCRIPTION
**Pull request by mkeeter**
See https://github.com/oxidecomputer/management-gateway-service/pull/316
and https://github.com/oxidecomputer/hubris/pull/1921 for details

This PR adds a new `humility hydrate` subcommand, which takes a "dry" task dump and staples it to a Hubris archive, creating a normal Hubris core file.

### CODE REVIEW

File: cmd/hydrate/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        data.copy_from_slice(&mem[offset..end]);
+        Ok(())
+    }
+}
+
+fn run(context: &mut ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = Args::try_parse_from(subargs)?;
+    let f = std::fs::File::open(&subargs.file)?;
+    let mut z = zip::ZipArchive::new(f)?;
```
Comment (labbott) [inline cmd/hydrate/src/lib.rs:107]: How well documented/stable is this format and how mad are we going to be if we need to change it in the future?

Comment (mkeeter) [reply cmd/hydrate/src/lib.rs:107]: I don't expect it to change much, because there's only so many ways to represent a bunch of bytes.  Still, I added `meta.json` containing a `version` value.

File: cmd/hydrate/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    // compare archive ID
+    if context.cli.archive.is_none() {
+        bail!("must provide an archive");
+    }
+    let archive = context.archive.as_ref().unwrap();
+    let expected_id = archive
+        .image_id()
+        .ok_or_else(|| anyhow!("missing image ID in archive"))?;
+    if archive_id != expected_id {
+        bail!("image ID mismatch: archive ID is {expected_id:02x?}");
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:193]: should we also include the actual archive ID in this error message?

File: cmd/hydrate/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility hydrate`
+//!
+//! `humility hydrate` combines a raw memory dump (obtained from MGS) and a
+//! Hubris archive, forming a proper Hubris dump.
+//!
+//! ```console
+//! $ humility -a build-grapefruit-image-default.zip hydrate hubris.dry.0
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:11]: this is probably not a big deal, but it might be nice if we actually stated somewhere explicitly that the archive is required for this command?

**Review by labbott — APPROVED** [review]

File: cmd/hydrate/src/lib.rs (hunk 1 above)
Comment (hawkw) [reply cmd/hydrate/src/lib.rs:107]: it occurs to me that if we want to make it a _little_ easier to evolve the format in the future, we could maybe stuff the paths to the various file entries we read out of the zip file in `meta.json`, so the format is a little bit more self-describing? i.e. instead of `z.by_name("TASK_INDEX")`, you would do `z.by_name(meta.task_index_path)` and so on. 

that way, if the paths change, we don't actually need to treat that as a totally new version that has to be handled separately. on the other hand, i'm not sure if changing these paths is actually a thing that will, like, ever reasonably happen, so maybe this isn't worth the effort, even if it's not that much effort.

File: cmd/hydrate/src/lib.rs (hunk 3 above)
Comment (hawkw) [reply cmd/hydrate/src/lib.rs:11]: also, maybe the docs should explicitly note that the result of this command is that a new core file for the dump is written to the filesystem in the current working directory? i realize that this is the same behavior as other dump commands, but it might be nice to write it down in the docs for this command so that it's obvious what it does...

File: cmd/hydrate/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    z.by_name("TIMESTAMP")?.read_to_string(&mut s)?;
+    let timestamp: u64 = s.trim().parse()?;
+
+    let mut bord = String::new();
+    z.by_name("BORD")?.read_to_string(&mut bord)?;
+    bord.pop(); // remove newline
+
+    let mut gitc = String::new();
+    z.by_name("GITC")?.read_to_string(&mut gitc)?;
+    gitc.pop(); // remove newline
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:137]: nit, take it or leave it: perhaps we ought to make this explicitly remove _only_ a newline, rather than "whatever the last character is"? i realize these are always expected to end with a newline, but it might be nice to be forward compatible and handle any potential future changes where they're _not_ newline-terminated, rather than just happily corrupting the commit SHA or board name if the newline ever goes away?

```suggestion
    let bord = bord.trim_end_matches('\n'); // remove newline

    let mut gitc = String::new();
    z.by_name("GITC")?.read_to_string(&mut gitc)?;
    let gitc = gitc.trim_end_matches('\n');; // remove newline
```

File: cmd/hydrate/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+    let vers = match z.by_name("VERS") {
+        Ok(mut v) => {
+            let mut s = String::new();
+            v.read_to_string(&mut s)?;
+            s.pop();
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:143]: similarly, perhaps we ought to `trim_end_matches` here?

File: cmd/hydrate/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    let meta: Meta = serde_json::from_str(&s)?;
+    if meta.version != 1 {
+        bail!(
+            "invalid version in `meta.json`: expected 1, got {}",
+            meta.version
+        );
+    }
+
+    let mut s = String::new();
+    z.by_name("TASK_INDEX")?.read_to_string(&mut s)?;
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:124]: nit: looks like the `ZipError::FileNotFound` variant from the `zip` crate doesn't include the filename, so perhaps we ought to do something like:
```suggestion
    z.by_name("TASK_INDEX")
      .context("invalid dump: missing TASK_INDEX")?
      .read_to_string(&mut s)
      .context("failed to read TASK_INDEX from dump")?;
```
or similar, here and for every other time we read a file?

if that's too wordy, we could also have a helper like
```rust
fn read_zip_entry(z: &mut ZipArchive, file: &str) -> anyhow::Result<String> {
    let mut s = String::new();
    z.by_name(file)
      .with_context(|| format!("invalid dump: missing {file}"))?
      .read_to_string(&mut s)
      .with_context(|| format!("failed to read {file} from dump"))?;
    Ok(s)
}
```
and just call that everywhere we read something from the dry dump?

File: cmd/hydrate/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    msg!("  {} memory regions:", mem.len());
+    for (k, v) in &mem {
+        msg!("    {k:#08x}: {} bytes", v.len());
+    }
+
+    // compare archive ID
+    if context.cli.archive.is_none() {
+        bail!("must provide an archive");
+    }
+    let archive = context.archive.as_ref().unwrap();
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:188]: nitpicky, feel free to ignore: why not golf these two down to just:
```suggestion
    let Some(archive) = context.archive.as_ref() else {
        bail!("must provide an archive");
    };
```

File: cmd/hydrate/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    for (i, b) in archive_id.iter_mut().enumerate() {
+        *b = u8::from_str_radix(&s[i * 2..][..2], 16)?;
+    }
+
+    let mut mem = BTreeMap::new();
+    let mem_files = z
+        .file_names()
+        .filter(|f| f.starts_with("0x"))
+        .map(|s| s.to_owned())
+        .collect::<Vec<_>>();
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:162]: ~~nit, not a big deal: we should be able to drop the `collect` here, and instead just `for` loop over the iterator directly, right?~~

EDIT: oh, I bet this is necessary because the iterator borrows the `ZipArchive`, but calling `ZipArchive::by_name` inside the loop mutably borrows it, so we can't do that. Maybe worth a comment here?

**Review by hawkw — APPROVED** [review]
Overall, the command looks good to me! I had a bunch of fairly minor little suggestions, but no major issues.

File: cmd/hydrate/src/lib.rs (hunk 2 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs]: Yup, done

File: cmd/hydrate/src/lib.rs (hunk 3 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs:11]: Good idea, done

File: cmd/hydrate/src/lib.rs (hunk 7 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs]: That's not _exactly_ equivalent, because `context.cli.archive` and `context.archive` are different things!  However, I'm not sure why they'd ever differ, so this is probably a fine suggestion; done

File: cmd/hydrate/src/lib.rs (hunk 1 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs:117]: I decided to move the metadata from scattered files in the ZIP archive into the JSON file (renamed to `dump.json`), which lets us deserialize with `serde` (and makes the questions about trailing newlines moot)

File: cmd/hydrate/src/lib.rs (hunk 6 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs]: This is somewhat moot because I moved metadata into the JSON file, but I added a few `with_context(...)` annotations elsewhere.

File: cmd/hydrate/src/lib.rs (hunk 7 above)
Comment (hawkw) [reply cmd/hydrate/src/lib.rs]: Oh! I can't read --- I thought they were both `context.archive`. Whoops!

File: cmd/hydrate/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+//!
+//! ```console
+//! $ humility -a build-grapefruit-image-default.zip hydrate hubris.dry.0
+//! ```
+//!
+//! The archive is required, because the raw memory dump does not contain debug
+//! information.
+//!
+//! By default, this writes to `hubris.core.TASK_NAME.N` (where `N` is the
+//! lowest available value); use `--out` to specify a different path name.
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:18]: lovely, thank you for adding both of these!

File: cmd/hydrate/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+#[clap(
+    name = "hydrate", about = env!("CARGO_PKG_DESCRIPTION"),
+    group = ArgGroup::new("target").multiple(false)
+)]
+struct Args {
+    /// Path to write the resulting dump
+    #[clap(short, long)]
+    out: Option<PathBuf>,
+
+    /// Path to the raw memory dump
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:39]: turbo-nit: we _may_ want to add
```suggestion
    #[clap(short, long, value_hint = clap::ValueHint::FilePath)]
    out: Option<PathBuf>,

    /// Path to the raw memory dump
    #[clap(value_hint = clap::ValueHint::FilePath)]
```
to generate nicer completions, but...i don't believe we're currently generating completions for humility anyway, so it may not matter...

File: cmd/hydrate/src/lib.rs (hunk 1 above)
Comment (hawkw) [reply cmd/hydrate/src/lib.rs:117]: lovely!

File: cmd/hydrate/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    #[derive(serde::Deserialize)]
+    struct DumpInfo {
+        format: u64,
+        task_index: u16,
+        crash_time: u64,
+        board_name: String,
+        git_commit: String,
+        archive_id: String,
+        fw_version: Option<String>,
+    }
```
Comment (hawkw) [inline cmd/hydrate/src/lib.rs:133]: part of me wonders if it might be worth having a shared definition of this structure that could also be used by MGS instead of the [ad-hoc JSON here](https://github.com/oxidecomputer/management-gateway-service/pull/316/files#diff-eecb3109d05f85699dc42276b1a71b8ec4041db954cce5ba3a1496c386685bf0R1583-R1589), but... i don't really know what crate we would put that in that it would make sense to have as a dependency of both MGS and humility. so, maybe it's not worth the effort... :woman_shrugging:

**Review by hawkw — APPROVED** [review]
latest changes look great!

File: cmd/hydrate/src/lib.rs (hunk 10 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs:39]: I think our use of Clap is weird enough that completions would be a challenge, so I'm going to punt on this for now.

File: cmd/hydrate/src/lib.rs (hunk 11 above)
Comment (mkeeter) [reply cmd/hydrate/src/lib.rs:133]: I think the dependency should probably go Humility → something in `management-gateway`, but agree that it's not worth the effort / coupling yet...
